h2o-genmodel/src/main/java/hex/genmodel/algos/gam/GamMojoModelBase.java

```diff
@@ -0,0 +1,175 @@
+package hex.genmodel.algos.gam;
+
+import hex.genmodel.ConverterFactoryProvidingModel;
+import hex.genmodel.GenModel;
+import hex.genmodel.MojoModel;
+import hex.genmodel.easy.CategoricalEncoder;
+import hex.genmodel.easy.EasyPredictModelWrapper;
+import hex.genmodel.easy.RowData;
+import hex.genmodel.easy.RowToRawDataConverter;
+import hex.genmodel.utils.ArrayUtils;
+import hex.genmodel.utils.DistributionFamily;
+import hex.genmodel.utils.LinkFunctionType;
+
+import java.util.Arrays;
+import java.util.Map;
+
+public abstract class GamMojoModelBase extends MojoModel implements ConverterFactoryProvidingModel {
+  String _link;
+  public LinkFunctionType _link_function;
+  boolean _useAllFactorLevels;
+  int _cats;
+  int[] _catNAFills;
+  int[] _catOffsets;
+  int _nums;
+  int _numsCenter;
+  double[] _numNAFills;
+  double[] _numNAFillsCenter;
+  boolean _meanImputation;
+  double[] _beta;
+  double[] _beta_no_center;
+  double[] _beta_center;
+  double[][] _beta_multinomial;
+  double[][] _beta_multinomial_no_center; // coefficients not centered for multinomial/ordinal
+  double[][] _beta_multinomial_center; // coefficients not centered for multinomial/ordinal
+  DistributionFamily _family;
+  String[] _gam_columns;
+  int _num_gam_columns;
+  int[] _bs;
+  int[] _num_knots;
+  double[][] _knots;
+  double[][][] _binvD;
+  double[][][] _zTranspose;
+  String[][] _gamColNames;  // expanded gam column names
+  String[][] _gamColNamesCenter;
+  String[] _names_no_centering; // column names of features with no centering
+  int _totFeatureSize; // number of predictors plus gam columns no centered
+  int _betaSizePerClass;
+  int _betaCenterSizePerClass;
+  double _tweedieLinkPower;
+  double[][] _basisVals; // store basis values array for each gam column
+  double[][] _hj;   // difference between knot values
+  int _numExpandedGamCols; // number of expanded gam columns
+  int _lastClass;
+  int _predictColumnNum; // Count number of columns that are found both in input data row and gam predictor columns
+  
+  public int getTotFeatureSize() { return _totFeatureSize;}
+  public String[] getNames() { return _names; }  // get column names centered
+  public int getPredictColumnNum() { return _predictColumnNum;}
+  
+  GamMojoModelBase(String[] columns, String[][] domains, String responseColumn) {
+    super(columns, domains, responseColumn);
+  }
+
+  public void setPredictColumNum(String[] inputColumnNames) {
+    _predictColumnNum = 0;
+    String[] colNames = _names; // contains predictor columns for Gam including expanded gam columns
+    for (String cName : colNames) {
+      if (Arrays.asList(inputColumnNames).contains(cName))
```

Comment:
can it happen that `_predictColumnNum != inputColumnNames.lenth - _names.length`? if not than this is unnecessary and can be done in predict

Reply:
@honza:  Yes, this can happen.  With GAM, gam_columns will be added to the training frame.  When you send me a training dataset, I will add the expanded gam_columns to the training frame.  If you want to do scoring with test dataset, it will not have the expanded gam_columns and hence inputColumnNames will be less than the _names that includes the predictors and expanded gam columns.

Reply:
this is MOJO we are talking about, so there is no train/test dataset, there is only the production dataset, that is the same structure as train, isnt that right?

Reply:
@honzasterba :  That is correct.  The user will give us a dataset that have the same predictor columns as in the train dataset.  Note that the gam_columns are built off the train dataset predictors that the user specified in gam_columns.  Here is the data flow:
1. input: train dataset,
2. add gam_columns: attach gam_columns that are derived from predictor columns in train dataset.

When we have scoring, we will do the same thing:
1. input: production/test dataset
2. add gam_columns: attach gam_columns that are derived from predictor columns in test dataset.

Is this clear?  Please let me know.